src/common/ipc.ts

```diff
@@ -2,94 +2,133 @@
 // https://www.electronjs.org/docs/api/ipc-main
 // https://www.electronjs.org/docs/api/ipc-renderer
 
-import { ipcMain, ipcRenderer, WebContents, webContents } from "electron"
+import { ipcMain, ipcRenderer, IpcRendererEvent, WebContents, webContents } from "electron"
 import logger from "../main/logger";
+import { getRandId } from "./utils";
 
 export type IpcChannel = string;
 
-export interface IpcHandleOpts {
-  timeout?: number;
+export enum IpcMode {
+  SYNC = "sync",
+  ASYNC = "async",
 }
 
-export interface IpcMessageHandler<T extends any[] = any> {
-  (...args: T): any;
+export interface IpcChannelRequest<A extends any[] = any> {
+  msgId: string;
+  args: A;
 }
 
-export interface IpcMessageOpts<A extends any[] = any> {
+export interface IpcChannelResponse<T extends any[] = any, E = any> {
+  msgId: string;
+  data?: T;
+  error?: E;
+}
+
+export interface IpcChannelInit {
+  channel: IpcChannel; // main <-> renderer communication channel name
+  mode?: IpcMode; // default: "async", use "sync" as last resort: https://www.electronjs.org/docs/api/ipc-renderer#ipcrenderersendsyncchannel-args
+  handle?: (...args: any[]) => any; // main-process message handler
+  autoBind?: boolean; // auto-bind message handler in main-process, default: false
+  timeout?: number; // timeout for waiting response from the sender
+  once?: boolean; // todo: add support
+}
+
+export function createIpcChannel({ autoBind = true, mode = IpcMode.ASYNC, timeout = 0, handle, channel }: IpcChannelInit) {
+  channel = `${mode}:${channel}`
```

Comment:
Shouldn't it be an error to given `channel === ""`?

Reply:
doesn't really matter in this case, but ok

Reply:
not critical imo, why not an empty string as a name?

Reply:
Just seems like a weird that a channel called `"sync:"` or `"async:"` could be made. Another possible solution could be 

```javascript
channel = `${mode}:${channel || "default"}`
```